server/hwapi/data_models/repository.py

```diff
@@ -107,3 +111,253 @@ def get_or_create(
+    db: Session,
+    devices_query: Query,
+    network_validator: device_validators.NetworkAdapterValidator,
+) -> models.Device | None:
+    """
+    Find a network wired device across the given list of devices that has the same
+    idenitifier, model name, and bus
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == network_validator.identifier.lower(),
+            models.Device.name == network_validator.model,
+            models.Device.bus == network_validator.bus,
+            models.Device.category.in_(
+                [DeviceCategory.NETWORK.value, DeviceCategory.OTHER]
+            ),
+        ),
+    ).first()
+
+
+def find_matching_wireless_device(
+    db: Session,
+    devices_query: Query,
+    wireless_validator: device_validators.WirelessAdapterValidator,
+) -> models.Device | None:
+    """
+    Find a network wireless device across the given list of devices that has the same
+    identifier and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == wireless_validator.identifier.lower(),
+            models.Device.name == wireless_validator.model,
+            models.Device.category.in_([DeviceCategory.WIRELESS, DeviceCategory.OTHER]),
+        ),
+    ).first()
+
+
+def find_matching_audio_device(
+    db: Session,
+    devices_query: Query,
+    audio_validator: device_validators.AudioValidator,
+) -> models.Device | None:
+    """
+    Find a network audio device across the given list of devices that has the same
+    identifier and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == audio_validator.identifier.lower(),
+            models.Device.name == audio_validator.model,
+            models.Device.category.in_([DeviceCategory.AUDIO, DeviceCategory.OTHER]),
+        ),
+    ).first()
+
+
+def find_matching_capture_device(
+    db: Session,
+    devices_query: Query,
+    capture_validator: device_validators.VideoCaptureValidator,
+) -> models.Device | None:
+    """
+    Find a network video capture device across the given list of devices that
+    has the same identifier and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == capture_validator.identifier.lower(),
+            models.Device.name == capture_validator.model,
+            models.Device.category.in_([DeviceCategory.CAPTURE, DeviceCategory.OTHER]),
+        ),
+    ).first()
+
+
+def find_matching_pci_device(
+    db: Session,
+    devices_query: Query,
+    pci_validator: device_validators.PCIPeripheralValidator,
+) -> models.Device | None:
+    """
+    Find an arbitary pci device across the given list of devices that has the same
+    PCI ID and model name
+    """
+    return devices_query.filter(
+        and_(
+            models.Device.identifier == pci_validator.pci_id.lower(),
+            models.Device.bus == BusType.pci,
```

Comment:
should `Device.bus` be renamed `Device.bus_type`?

Reply:
hm, since we don't store bus as an object, I don't see a reason to do it. shall we change it?

Reply:
No, you don't need to reflect the type name in a variable or a property.